kvserver/rangefeed: wip

wenyihu6 opened the pull request:
hope this is the last integration branch 

epic: none

wenyihu6 committed:
to be squashed separated out for easier code review

# Conflicts:
#	pkg/kv/kvserver/rangefeed/processor_test.go
#	pkg/kv/kvserver/rangefeed/registry_helpers_test.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/processor_helpers_test.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/processor_test.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/processor_test.go

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration tests

This patch adds more test cases for unbufferedRegistration.

Closed: #126560
Release note: none

# Conflicts:
#	pkg/kv/kvserver/rangefeed/registry_test.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/registry_test.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/registry_test.go
#	pkg/kv/kvserver/rangefeed/scheduled_processor.go
#	pkg/kv/kvserver/rangefeed/sender_helper_test.go

wenyihu6 committed:
to be squashed separated out for easier code review

# Conflicts:
#	pkg/kv/kvserver/rangefeed/registry_test_helper.go

wenyihu6 committed:
kvserver/rangefeed: add unbuffered registration

This patch adds unbufferedRegistration.

UnbufferedRegistration is like BufferedRangefeed but uses BufferedStream to
buffer live raft updates instead of a using buf channel and having a
dedicated per-range per-registration goroutine to volley events to underlying
grpc stream. Instead, there is only one BufferedStream for each incoming
node.MuxRangefeed rpc call. BufferedStream is responsible for buffering and
sending its updates to the underlying grpc stream in a dedicated goroutine
O(node).

Note that BufferedStreamSender is still left unimplemented. This commit doesn't
add any tests for it yet. More tests will be added in future commits of this PR.

Part of: #129814
Release note: none

# Conflicts:
#	pkg/kv/kvserver/rangefeed/scheduled_processor.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/scheduled_processor.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/scheduled_processor.go

wenyihu6 committed:
kvserver/rangefeed: refactor registry_helpers_test to facilitate future tests

This patch refactors helper functions in registry_helpers_test.

Epic: none
Release note: none

wenyihu6 committed:
kvserver/rangefeed: add node level buffered sender

This patch is the last step for reducing long-running O(ranges) goroutines in
kvserver/rangefeed. It changes the BufferedSender to use a queue to buffer
events before forwarding them to underlying grpc stream.

Closed: #129813
Release note: A new cluster setting
`kv.rangefeed.buffered_stream_sender.enabled` can now be used to allow rangefeed
to use buffered sender for all rangefeed feeds instead of buffering events
separately per client per range.

# Conflicts:
#	pkg/kv/kvserver/rangefeed/buffered_sender.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/buffered_sender.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/buffered_sender.go

# Conflicts:
#	pkg/kv/kvserver/rangefeed/buffered_sender.go
#	pkg/kv/kvserver/rangefeed/buffered_stream.go
#	pkg/kv/kvserver/rangefeed/stream.go
#	pkg/kv/kvserver/rangefeed/unbuffered_sender.go
#	pkg/kv/kvserver/rangefeed/unbuffered_sender_test.go
#	pkg/server/node.go

wenyihu6 committed:
kvserver/rangefeed: add event queue

This patch adds a new data structure eventQueue which is like a queue but uses a
fixed size for the chunked linked list. Each chunk has a fixed size of 4096
elements. This implementation uses sync.Pool to reduce the number of
allocations.

pushBack, popFront, len run in constant time. removeAll runs in linear time with
respect to the number of elements in the queue. This structure is not safe for
concurrent use.

This is for future commits to include the queue in the BufferedSender to buffer
events at the node level.

Part of: #129813
Release note: none

wenyihu6 committed:
kvserver/rangefeed: remove context from kvpb.RangeFeedEventSink

Previously, `node.MuxRangefeed` created a child context for each rangefeed
request, storing it in the stream interface to allow the node level to be able
to shut down registration goroutines. This patch simplifies the approach by
passing the stream context directly to `p.Register`, eliminating the need to
store context in `streamSink` or return context via the interface. So this patch
also removes context from `kvpb.RangeFeedEventSink`.

Epic: none
Release note: none

stevendanna committed:
kv/kvserver: remove redundant cancel func

Previously, UnbufferedSender maintained a list of cancel functions
that were canceled when processing an error. The goal of this
cancellation was to tear down running resource in the registration.

On error, something like the following would happen:

UnbufferedSender.SendBufferedError(err) would call cancel.  This
cancellation would cancel the context inherited by the registration's
runOutputLoop, causing it to exit. The registration would then call
disconnect, cancelling additional contexts, that also seem to be
intended to control the output loop, which is no longer
running. Disconnect would then send an error back to UnbufferedSender,
which would ignore it to prevent a duplicate sending of errors.

There are other error paths, where some of the components describe
above, aren't unnecessary, but, hopefully you can see that this is
ripe for some refactoring.

Here, we take a small step that we believe will also pave the way for
a clearer implementation of the BufferredSender.

Now, all error paths, start by calling
registration.Disconnect(err). The registration, cancels the context
controlling the runOutputLoop and sends an error to the stream, which
the stream then sends to the client.

There are some paths were Disconnect may be called twice still.

This simplification is made possible by one bit of indirection: The
processor now adds the registration to a map that lives on the stream
manager. This map will eventually also store any deferred cleanups
that the process requires when that stream is shutting down.

Note, we considered passing the registrations cleanup function back up
the call stack from register and may still explore that in the
future. However, it is not without challenge. Namely, the second we've
added the registration to the processor, it is possible that an error
is delivered from the processor that the stream manager would not yet
know how to handle.

We expect further naming and organizational improvements over the
coming commits.

Epic: none
Release note: None

# Conflicts:
#	pkg/kv/kvserver/rangefeed/processor_test.go
#	pkg/server/node.go

stevendanna committed:
kvserver/rangefeed: process unregister requests asyncronously

After the output loop completes, 2 cleanup steps are performed:

1) The registration is removed from the processors registry
2) The process is potentially stopped and removed from the replica.

Before this change, both of these happened syncronously after the
output loop finished.

With this change, step (1) happens asyncronously. To facilitate this,
an overflow mechanism is provided. This overflow mechanism potentially
allocates. Note that we expect that the number of requests is
relatively small and should be O(rangefeeds_on_range) so hopefully
this mechanism won't be used often.

Step (2) is now handled by the processor itself. After processing an
unregister request, if the set of registrations falls to zero, we
enqueue a Stop event for ourselves. Then, when processing the Stop, we
unregister ourselves from the replica.

Note that this may look like a small semantics change since the
previous unregister callback called Stop() which processes all events.
However, note that a Stopped event is processed after all other
events, so any events in the queue at the point of processing the
unregistration that enqueued the stop will be processed.

The motivation for this change is to eventually allow cleanup step (1)
to be run as part of registration.disconnect(), which needs to be
non-blocking. This is desired for a future change in which there is
not a dedicated goroutine to perform this cleanup.

Epic: none
Release note: None

stevendanna committed:
kvserver/rangefeed: scan all registrations overflow

Rather than a dynamic overflow queue, we now set a flag indicating
that an overflow happened and then scan all registrations in the case
of overflow.

Release note: None
Epic: none

stevendanna committed:
kvserver/rangefeed: run unregister as part of disconnect

For the ScheduledProcess, we now enqueue the unregister as part of the
first disconnect call.

Epic: none
Release note: None

# Conflicts:
#	pkg/kv/kvserver/rangefeed/buffered_registration.go
#	pkg/kv/kvserver/rangefeed/processor.go
#	pkg/kv/kvserver/rangefeed/registry.go
#	pkg/kv/kvserver/rangefeed/registry_test.go
#	pkg/kv/kvserver/rangefeed/scheduled_processor.go

wenyihu6 committed:
refactored

wenyihu6 committed:
clean up

wenyihu6 committed:
a

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/133133)
<!-- Reviewable:end -->

wenyihu6 committed:
remove rangeID from DisconnectStream

wenyihu6 committed:
fix more tests

wenyihu6 committed:
refactored runOutputLoop

wenyihu6 committed:
shouldn't be needed any more

wenyihu6 committed:
additional cleanup

stevendanna commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -51,54 +56,130 @@ type BufferedSender struct {
+// and release it if the return error is non-nil. Note that it is safe to send
+// error events without being blocked for too long.
+func (bs *BufferedSender) send(ev *kvpb.MuxRangeFeedEvent, alloc *SharedBudgetAllocation) error {
+	// Make sure this is from catch up scan.
```
Apologies, I know we've gone back and forth on this. This code should not need to know anything about catchup scan so I'm a little concerned by this comment. Let's pair on this a bit today.

wenyihu6 replied:
Paired on this today and got a shape we were happy with.

stevendanna commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -51,54 +56,130 @@ type BufferedSender struct {
+	bs.queueMu.Lock()
+	defer bs.queueMu.Unlock()
+	if bs.queueMu.stopped {
+		log.Errorf(context.Background(), "stream sender is stopped")
```
The caller can log this error if they care to.

wenyihu6 replied:
Removed.

stevendanna commented on pkg/kv/kvserver/rangefeed/event_queue.go:
```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 The Cockroach Authors.
```
:D Hopefully later today we can drop this from this PR.

stevendanna commented on pkg/kv/kvserver/rangefeed/processor_helpers_test.go:
```diff
@@ -23,6 +23,31 @@ import (
+		// In practise, StreamMuxer is responsible for invoking cleanup in the
+		// background. For simplicity, we call it directly here.
+		go tb.cleanup()
+		time.Sleep(10 * time.Millisecond)
```
What's this sleep for?

wenyihu6 replied:
iirc it was a hacky way to allow the goroutine running this code to be put to sleep and preempted, giving tb.cleanup a chance to take place

stevendanna commented on pkg/kv/kvserver/rangefeed/scheduled_processor.go:
```diff
@@ -304,21 +320,31 @@ func (p *ScheduledProcessor) Register(
-	if _, ok := stream.(BufferedStream); ok {
-		log.Fatalf(context.Background(),
-			"unimplemented: unbuffered registrations for rangefeed, see #126560")
+	bufferedStream, isBufferedStream := stream.(BufferedStream)
+	if isBufferedStream {
+		r = newUnbufferedRegistration(streamCtx,
+			span.AsRawSpanWithNoLocals(), startTS, catchUpIter, withDiff, withFiltering, withOmitRemote,
+			p.Config.EventChanCap, p.Metrics, bufferedStream, disconnectFn,
+			func(ctx context.Context, r registration) {
+				p.unregisterClientAsync(r)
+			})
 	} else {
 		r = newBufferedRegistration(
+			streamCtx,
 			span.AsRawSpanWithNoLocals(), startTS, catchUpIter, withDiff, withFiltering, withOmitRemote,
 			p.Config.EventChanCap, blockWhenFull, p.Metrics, stream, disconnectFn,
+			func(ctx context.Context, r registration) {
+				p.unregisterClientAsync(r)
+			},
 		)
 	}
```
Let's not change it now, but as a potential follow up, a way to do this without the type switch would be for stream to have a method like SendIsNonBlocking().

wenyihu6 replied:
I have some bad news: we can’t completely remove BufferedStream as we had done this morning because we still rely on the two different structs to determine which type of rangefeed registration to create right here. What you proposed here would help us solve this. 

But I’m also not sure whether adding a method like SendIsNonBlocking to the Stream interface is a good idea. Stream interface is used across so many test implementations. And it seems nicer to keep this interface minimal. Perhaps we could discuss this further and address this as a follow up.

stevendanna commented on pkg/kv/kvserver/rangefeed/stream.go:
```diff
@@ -61,22 +56,22 @@ func (s *PerRangeEventSink) SendUnbuffered(event *kvpb.RangeFeedEvent) error {
 	})
-	s.wrapped.SendBufferedError(ev)
+	if err := s.wrapped.send(ev, nil); err != nil {
+		log.Errorf(context.Background(),
```
It'll be annoying, but I think we should look into threading a context through some of our send methods as a follow up. We seem to both want to log and need a context for releasing allocations. It would be much nicer for these log messages to have proper log tags which they won't have with context.Background().

wenyihu6 replied:
Good idea, I will file an issue as a follow up on this one.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
I’m not sure if this is a valid concern, but I was worried about publishCatchUpBuffer being called concurrently with discardCatchUpBuf.  Both functions use channels, but they are both reading from the same channel. My concern was whether discardCatchUpBuf could drop some events while publishCatchUpBuffer is still running and sending more events, leading to dropped events. With the current setup, this shouldn’t happen since both functions are in the same goroutine. However, it might become an issue if we move discardCatchUpBuf to drainAllocations, which can be called from other goroutines.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
Discussed in the morning - it is possible and we are discarding during p.Register if runOutputLoop fails to start.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/buffered_sender.go:
```diff
@@ -51,54 +56,130 @@ type BufferedSender struct {
+		buffer  *eventQueue
+	}
+
+	// Unblocking channel to notify the BufferedSender.run goroutine that there
+	// are events to send.
+	notifyDataC chan struct{}
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/scheduled_processor.go:
```diff
@@ -344,37 +370,46 @@ func (p *ScheduledProcessor) Register(
-			r.disconnect(kvpb.NewError(err))
-			p.reg.Unregister(ctx, r)
+			r.Disconnect(kvpb.NewError(err))
+			// Normally, ubr.runOutputLoop is responsible for draining catch up
+			// buffer. If it fails to start, we should drain it here. Double check if
+			// runOutputLoop is guaranteed to be invoked if err == nil here.
+			if ubr, ok := r.(*unbufferedRegistration); ok {
+				ubr.discardCatchUpBuffer(ctx)
+			}
```
Not done yet. But happy to add this piece in if we still want it.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream.go:
```diff
@@ -61,22 +56,22 @@ func (s *PerRangeEventSink) SendUnbuffered(event *kvpb.RangeFeedEvent) error {
-	s.wrapped.SendBufferedError(ev)
+	if err := s.wrapped.send(ev, nil); err != nil {
+		log.Errorf(context.Background(),
+			"failed to send rangefeed completion error back to client due to broken stream: %v", err)
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+
+// StreamManager manages one or more streams.
+//
+// Implemented by BufferedSender and StreamManager.
```
Done

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+//
+// Implemented by BufferedSender and StreamManager.
+type StreamManager struct {
+	// taskCancel is a function to cancel UnbufferedSender.run spawned in the
+	// background. It is called by UnbufferedSender.Stop. It is expected to be
+	// called after UnbufferedSender.Start.
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+	// called after UnbufferedSender.Start.
+	taskCancel context.CancelFunc
+
+	// wg is used to coordinate async tasks spawned by UnbufferedSender.
+	// Currently, there is only one task spawned by UnbufferedSender.Start
+	// (UnbufferedSender.run).
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+	// (UnbufferedSender.run).
+	wg sync.WaitGroup
+
+	// errCh is used to signal errors from UnbufferedSender.run back to the
+	// caller. If non-empty, the UnbufferedSender.run is finished and error should
+	// be handled. Note that it is possible for UnbufferedSender.run to be
+	// finished without sending an error to errCh. Other goroutines are expected
+	// to receive the same shutdown signal in this case and handle error
+	// appropriately.
```
I removed it since it could be confusing, but the intent was to justify why we don’t return errors when the context is cancelled or when the stopper is quiescing. node.MuxRangefeed should be shutting down since it would be receiving the context cancellation and stopper quiescing.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+
+func (sm *StreamManager) DisconnectStream(streamID int64, err *kvpb.Error) {
+	if err == nil {
+		log.Fatalf(context.Background(), "unexpected: SendWithoutBlocking called with non-error event")
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+	}
+}
+
+// lock in streamManager.m
+// lock in registration.IsDisconnected
+
+// We should never add a stream if the stream has disconnected and OnError had been called.
+// Nothing would trigger a cleanup.
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/stream_manager.go:
```diff
@@ -0,0 +1,167 @@
+
+func (sm *StreamManager) Error() chan error {
+	if sm.errCh == nil {
+		log.Fatalf(context.Background(), "BufferedSender.Error called before BufferedSender.Start")
```
Done.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
```diff
@@ -0,0 +1,497 @@
+	"github.com/cockroachdb/errors"
+)
+
+// unbufferedRegistration is similar to bufferedRangefeed but uses
+// BufferedStream to buffer live raft updates instead of a using buf channel and
+// having a dedicated per-range per-registration goroutine to volley events to
+// underlying grpc stream. Instead, there is only one BufferedStream for each
+// incoming node.MuxRangefeed rpc call. BufferedStream is responsible for
+// buffering and sending its updates to the underlying grpc stream in a
+// dedicated goroutine O(node).
```
Addressed. Paired on this earlier.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
Removed.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
Addressed, paired on this earlier.

wenyihu6 commented on pkg/kv/kvserver/rangefeed/unbuffered_registration.go:
Removed.

wenyihu6 commented:
@stevendanna I've addressed your comments on our **_final_** integration branch https://github.com/cockroachdb/cockroach/pull/125872. Lmk if you have any further feedback on the PR.

stevendanna commented:
Closing this one.
